pydal/adapters/base.py

```diff
@@ -618,55 +623,71 @@ def _select_wcols(self, query, fields, left=False, join=False,
         #: build joins (inner, left outer) and table names
         if join:
             (
+                # FIXME? ijoin_tables is never used
                 ijoin_tables, ijoin_on, itables_to_merge, ijoin_on_tables,
-                iimportant_tablenames, iexcluded
-            ) = self._build_joins_for_select(tablenames, join)
+                iimportant_tablenames, iexcluded, itablemap
+            ) = self._build_joins_for_select(tablemap, join)
+            tablemap = merge_tablemaps(tablemap, itablemap)
         if left:
             (
                 join_tables, join_on, tables_to_merge, join_on_tables,
-                important_tablenames, excluded
-            ) = self._build_joins_for_select(tablenames, left)
+                important_tablenames, excluded, jtablemap
+            ) = self._build_joins_for_select(tablemap, left)
+            tablemap = merge_tablemaps(tablemap, jtablemap)
+        current_scope = outer_scoped + list(tablemap)
+        query_env = dict(current_scope=current_scope,
+            parent_scope=outer_scoped)
+        #: prepare columns and expand fields
+        colnames = [self._colexpand(x, query_env) for x in fields]
+        sql_fields = ', '.join(self._geoexpand(x, query_env) for x in fields)
+        table_alias = lambda name: tablemap[name].query_name(outer_scoped)[0]
         if join and not left:
             cross_joins = iexcluded + list(itables_to_merge)
-            sql_t = '%s' % self.table_alias(cross_joins[0])
-            for t in cross_joins[1:]:
-                sql_t += ' %s' % self.dialect.cross_join(self.table_alias(t))
-            for t in ijoin_on:
-                sql_t += ' %s' % self.dialect.join(t)
+            tokens = [table_alias(cross_joins[0])]
+            tokens += [self.dialect.cross_join(table_alias(t), query_env)
+                    for t in cross_joins[1:]]
+            tokens += [self.dialect.join(t, query_env) for t in ijoin_on]
+            sql_t = ' '.join(tokens)
         elif not join and left:
             cross_joins = excluded + list(tables_to_merge)
-            sql_t = '%s' % self.table_alias(cross_joins[0])
-            for t in cross_joins[1:]:
-                sql_t += ' %s' % self.dialect.cross_join(self.table_alias(t))
+            tokens = [table_alias(cross_joins[0])]
+            tokens += [self.dialect.cross_join(table_alias(t), query_env)
+                    for t in cross_joins[1:]]
+            # FIXME: WTF? This is not correct syntax at least on PostgreSQL
```

Comment:
`join_tables` gets expanded below the comment into something like `LEFT JOIN tab1, tab2, tab3, ...`. MySQL and PostgreSQL both raise syntax error on that but SQLite seems to accept it. At the very least, the expansion should be handled by the dialect, not `SQLAdapter._select_wcols()`.